--- cobbler/utils.py.orig
+++ cobbler/utils.py
@@ -941,7 +941,9 @@
       tokens = rest.split(" ")
       for t in tokens:
          try:
-             return (make,float(t))
+             match = re.match('^\d+(?:\.\d+)?', t)
+             if match:
+                return (make, float(match.group(0)))
          except ValueError, ve:
              pass
       raise CX("failed to detect local OS version from /etc/redhat-release")
