diff -rupN cobbler-2.0.11.old/cobbler/action_power.py cobbler-2.0.11/cobbler/action_power.py
--- cobbler-2.0.11.old/cobbler/action_power.py	2011-04-20 08:40:48.000000000 -0400
+++ cobbler-2.0.11/cobbler/action_power.py	2014-07-07 10:09:19.361118180 -0400
@@ -30,6 +30,7 @@ import os
 import os.path
 import traceback
 import time
+import re

 import utils
 import func_utils
@@ -108,8 +109,22 @@ class PowerTool:
         # Try the power command 5 times before giving up.
         # Some power switches are flakey
         for x in range(0,5):
-            rc = utils.subprocess_call(self.logger, cmd, shell=False)
+            output, rc = utils.subprocess_sp(self.logger, cmd, shell=False)
             if rc == 0:
+                # If the desired state is actually a query for the status
+                # return different information than command return code
+                if desired_state == 'status':
+                    match = re.match('(^Status:\s)(ON|OFF)', output)
+                    if not match:
+                        match = re.match('(.*Chassis power\s*(=|is)\s*)(On|Off).*', output, re.IGNORECASE)
+                    if match:
+                        power_status = match.groups()[2]
+                        if power_status.upper() == 'ON':
+                            return True
+                        else:
+                            return False
+                    utils.die(self.logger,"command succeeded (rc=%s), but output ('%s') was not understood" % (rc, output))
+                    return None
                 break
             else:
                 time.sleep(2)
diff -rupN cobbler-2.0.11.old/cobbler/api.py cobbler-2.0.11/cobbler/api.py
--- cobbler-2.0.11.old/cobbler/api.py	2014-07-07 09:41:53.412830237 -0400
+++ cobbler-2.0.11/cobbler/api.py	2014-07-07 09:42:26.844060269 -0400
@@ -820,6 +820,15 @@ class BootAPI:
         time.sleep(5)
         return self.power_on(system, user, password, logger=logger)

+    def power_status(self, system, user=None, password=None, logger=None):
+        """
+        Returns the power status for a system that has power management configured.
+
+        @return: 0  the system is powered on, False if it's not or None on error
+        """
+        return action_power.PowerTool(self._config, system, self, user, password, logger = logger).power("status")
+
+
     # ==========================================================================

     def clear_logs(self, system, logger=None):
diff -rupN cobbler-2.0.11.old/cobbler/cli.py cobbler-2.0.11/cobbler/cli.py
--- cobbler-2.0.11.old/cobbler/cli.py	2014-07-07 09:41:53.412830237 -0400
+++ cobbler-2.0.11/cobbler/cli.py	2014-07-07 09:42:26.844060269 -0400
@@ -39,7 +39,7 @@ import item_image
 OBJECT_ACTIONS   = {
    "distro"  : "add copy edit find list remove rename report".split(" "),
    "profile" : "add copy dumpvars edit find getks list remove rename report".split(" "),
-   "system"  : "add copy dumpvars edit find getks list remove rename report poweron poweroff reboot".split(" "),
+   "system"  : "add copy dumpvars edit find getks list remove rename report poweron poweroff powerstatus reboot".split(" "),
    "image"   : "add copy edit find list remove rename report".split(" "),
    "repo"    : "add copy edit find list remove rename report".split(" ")
 } 
@@ -277,7 +277,7 @@ class BootCLI:
             keys.sort()
             for x in keys:
                print "%s : %s" % (x, data[x])
-        elif object_action in [ "poweron", "poweroff", "reboot" ]:
+        elif object_action in [ "poweron", "poweroff", "powerstatus", "reboot" ]:
             power={}
             power["power"] = object_action.replace("power","")
             power["systems"] = [options.name]
diff -rupN cobbler-2.0.11.old/cobbler/remote.py cobbler-2.0.11/cobbler/remote.py
--- cobbler-2.0.11.old/cobbler/remote.py	2014-07-07 09:41:53.428830347 -0400
+++ cobbler-2.0.11/cobbler/remote.py	2014-07-07 09:42:26.845060276 -0400
@@ -1711,7 +1711,7 @@ class CobblerXMLRPCInterface:
         """
         Internal implementation used by background_power, do not call
         directly if possible.  
-        Allows poweron/poweroff/reboot of a system specified by object_id.
+        Allows poweron/poweroff/powerstatus/reboot of a system specified by object_id.
         """
         obj = self.__get_object(object_id)
         self.check_access(token, "power_system", obj)
@@ -1719,10 +1719,12 @@ class CobblerXMLRPCInterface:
             rc=self.api.power_on(obj, user=None, password=None, logger=logger)
         elif power=="off":
             rc=self.api.power_off(obj, user=None, password=None, logger=logger)
+        elif power=="status":
+            rc=self.api.power_status(obj, user=None, password=None, logger=logger)
         elif power=="reboot":
             rc=self.api.reboot(obj, user=None, password=None, logger=logger)
         else:
-            utils.die(self.logger, "invalid power mode '%s', expected on/off/reboot" % power)
+            utils.die(self.logger, "invalid power mode '%s', expected on/off/status/reboot" % power)
         return rc

     def clear_system_logs(self, object_id, token=None, logger=None):
diff -rupN cobbler-2.0.11.old/templates/power_ipmilan.template cobbler-2.0.11/templates/power_ipmilan.template
--- cobbler-2.0.11.old/templates/power_ipmilan.template	2011-04-20 08:40:48.000000000 -0400
+++ cobbler-2.0.11/templates/power_ipmilan.template	2014-07-07 09:42:50.837225387 -0400
@@ -1,2 +1 @@
-#use power_id to pass in additional options like -P, Use Lanplus
-fence_ipmilan -i "$power_address" $power_id -l "$power_user" -p "$power_pass" -o "$power_mode"
+fence_ipmilan -a "$power_address" $power_id -l "$power_user" -p "$power_pass" -o "$power_mode"
