diff -rupN cobbler-2.0.11.old/cobbler/api.py cobbler-2.0.11/cobbler/api.py
--- cobbler-2.0.11.old/cobbler/api.py	2015-04-22 14:33:29.451706375 -0400
+++ cobbler-2.0.11/cobbler/api.py	2015-04-22 14:35:04.563387830 -0400
@@ -179,11 +179,9 @@ class BootAPI:
         API instance, regardless of the serializer type.
         """
         if not os.path.exists("/var/lib/cobbler/.mtime"):
-            old = os.umask(0x777)
-            fd = open("/var/lib/cobbler/.mtime","w")
-            fd.write("0")
-            fd.close()
-            os.umask(old)
+            fd = os.open("/var/lib/cobbler/.mtime", os.O_CREAT|os.O_RDWR, 0200)
+            os.write(fd, "0")
+            os.close(fd)
             return 0
         fd = open("/var/lib/cobbler/.mtime")
         data = fd.read().strip()
diff -rupN cobbler-2.0.11.old/cobbler/cobblerd.py cobbler-2.0.11/cobbler/cobblerd.py
--- cobbler-2.0.11.old/cobbler/cobblerd.py	2015-04-22 14:33:29.445706332 -0400
+++ cobbler-2.0.11/cobbler/cobblerd.py	2015-04-22 14:37:29.421426143 -0400
@@ -58,17 +58,14 @@ def regen_ss_file():
     data = fd.read(512)
     fd.close()
     if not os.path.isfile(ssfile):
-        um = os.umask(int('0027',16))
-        fd = open(ssfile,"w+")
-        fd.write(binascii.hexlify(data))
-        fd.close()
-        os.umask(um)
-        utils.os_system("chmod 700 /var/lib/cobbler/web.ss")
+        fd = os.open(ssfile, os.O_CREAT|os.O_RDWR, 0600)
+        os.write(fd, binascii.hexlify(data))
+        os.close(fd)
         utils.os_system("chown apache /var/lib/cobbler/web.ss")
     else:
-        fd = open(ssfile,"w+")
-        fd.write(binascii.hexlify(data))
-        fd.close()
+        fd = os.open(ssfile, os.O_CREAT|os.O_RDWR, 0600)
+        os.write(fd, binascii.hexlify(data))
+        os.close(fd)
 
     return 1
 
diff -rupN cobbler-2.0.11.old/cobbler/serializer.py cobbler-2.0.11/cobbler/serializer.py
--- cobbler-2.0.11.old/cobbler/serializer.py	2015-04-22 14:33:29.447706346 -0400
+++ cobbler-2.0.11/cobbler/serializer.py	2015-04-22 14:38:42.756951796 -0400
@@ -64,11 +64,9 @@ def __release_lock(with_changes=False):
         # this file is used to know when the last config change
         # was made -- allowing the API to work more smoothly without
         # a lot of unneccessary reloads.  
-        old = os.umask(0x777)
-        fd = open("/var/lib/cobbler/.mtime","w")
-        fd.write("%f" % time.time())
-        fd.close()
-        os.umask(old)
+        fd = os.open("/var/lib/cobbler/.mtime", os.O_CREAT|os.O_RDWR, 0200)
+        os.write(fd, "%f" % time.time())
+        os.close(fd)
     if LOCK_ENABLED:
         LOCK_HANDLE = open("/var/lib/cobbler/lock","r")
         fcntl.flock(LOCK_HANDLE.fileno(), fcntl.LOCK_UN)
