diff -rupN cobbler-2.0.11.old/cobbler/codes.py cobbler-2.0.11/cobbler/codes.py
--- cobbler-2.0.11.old/cobbler/codes.py	2014-07-14 09:37:29.366466959 -0400
+++ cobbler-2.0.11/cobbler/codes.py	2014-07-14 09:41:35.536247306 -0400
@@ -41,7 +41,7 @@ VALID_OS_BREEDS = [
 ]
 
 VALID_OS_VERSIONS = {
-    "redhat"  : [ "rhel2.1", "rhel3", "rhel4", "rhel5", "rhel6", "fedora5", "fedora6", "fedora7", "fedora8", "fedora9", "fedora10", "fedora11", "fedora12", "fedora13", "fedora14", "fedora15", "generic24", "generic26", "virtio26", "other" ],
+    "redhat"  : [ "rhel2.1", "rhel3", "rhel4", "rhel5", "rhel6", "rhel7", "fedora5", "fedora6", "fedora7", "fedora8", "fedora9", "fedora10", "fedora11", "fedora12", "fedora13", "fedora14", "fedora15", "fedora16", "fedora17", "fedora18", "fedora19", "fedora20", "generic24", "generic26", "virtio26", "other" ],
     "suse"    : [ "sles10", "generic24", "generic26", "virtio26", "other" ],
     "debian"  : [ "etch", "lenny", "squeeze", "sid", "stable", "testing", "unstable", "generic24", "generic26", "other" ],
     "ubuntu"  : [ "dapper", "hardy", "intrepid", "jaunty" ],
