diff -rupN authn_spacewalk.py.orig authn_spacewalk.py
--- cobbler-2.0.11-orig/cobbler/modules/authn_spacewalk.py     2015-06-09 09:14:26.503449862 -0400
+++ cobbler-2.0.11/cobbler/modules/authn_spacewalk.py  2015-06-09 09:14:41.573214329 -0400
@@ -77,7 +77,7 @@ def authenticate(api_handle,username,pas
     if server == "xmlrpc.rhn.redhat.com":
         return False # emergency fail, don't bother RHN!

-    spacewalk_url = "https://%s/rpc/api" % server
+    spacewalk_url = "http://%s/rpc/api" % server

     client = xmlrpclib.Server(spacewalk_url, verbose=0)

