diff -up ./cobbler/codes.py.org ./cobbler/codes.py
--- ./cobbler/codes.py.org	2014-09-30 15:54:10.676556504 +0200
+++ ./cobbler/codes.py	2014-09-30 15:54:41.273650788 +0200
@@ -41,7 +41,7 @@ VALID_OS_BREEDS = [
 ]

 VALID_OS_VERSIONS = {
-    "redhat"  : [ "rhel2.1", "rhel3", "rhel4", "rhel5", "rhel6", "rhel7", "fedora5", "fedora6", "fedora7", "fedora8", "fedora9", "fedora10", "fedora11", "fedora12", "fedora13", "fedora14", "fedora15", "fedora16", "fedora17", "fedora18", "fedora19", "fedora20", "generic24", "generic26", "virtio26", "other" ],
+    "redhat"  : [ "rhel2.1", "rhel3", "rhel4", "rhel5", "rhel6", "rhel7", "fedora18", "generic24", "generic26", "virtio26", "other" ],
     "suse"    : [ "sles10", "generic24", "generic26", "virtio26", "other" ],
     "debian"  : [ "etch", "lenny", "squeeze", "sid", "stable", "testing", "unstable", "generic24", "generic26", "other" ],
     "ubuntu"  : [ "dapper", "hardy", "intrepid", "jaunty" ],
