diff -rupN cobbler-2.0.11-old2/koan/app.py cobbler-2.0.11/koan/app.py
--- cobbler-2.0.11-old2/koan/app.py	2013-12-12 17:32:03.506122558 -0500
+++ cobbler-2.0.11/koan/app.py	2013-12-12 17:58:14.016155914 -0500
@@ -189,6 +189,19 @@ def main():
     p.add_option("", "--qemu-disk-type",
                  dest="qemu_disk_type",
                  help="when used with --virt_type=qemu, add select of disk drivers: ide,scsi,virtio")
+    p.add_option("", "--qemu-net-type",
+                 dest="qemu_net_type",
+                 help="when used with --virt_type=qemu, select type of network device to use: e1000, ne2k_pci, pcnet, rtl8139, virtio")
+    p.add_option("", "--wait",
+                 dest="wait", type='int', default=0,    # default to 0 for koan backwards compatibility
+                 help="pass the --wait=<INT> argument to virt-install")
+    p.add_option("", "--cpu",
+                 dest="cpu",
+                 help="pass the --cpu argument to virt-install")
+    p.add_option("", "--noreboot",
+                 dest="noreboot", default=False,        # default to False for koan backwards compatibility
+                 action="store_true",
+                 help="pass the --noreboot argument to virt-install")
 
     (options, args) = p.parse_args()
 
@@ -218,6 +231,10 @@ def main():
         k.embed_kickstart     = options.embed_kickstart
         k.virt_auto_boot      = options.virt_auto_boot
         k.qemu_disk_type      = options.qemu_disk_type       
+        k.qemu_net_type       = options.qemu_net_type
+        k.virtinstall_cpu     = options.cpu
+        k.virtinstall_wait    = options.wait
+        k.virtinstall_noreboot= options.noreboot
 
         if options.virt_name is not None:
             k.virt_name          = options.virt_name
@@ -274,6 +291,10 @@ class Koan:
         self.force_path        = None
         self.qemu_disk_type    = None
         self.virt_auto_boot    = None
+        self.qemu_net_type     = None
+        self.virtinstall_cpu   = None
+        self.virtinstall_wait  = None
+        self.virtinstall_noreboot = None
 
         # This option adds the --copy-default argument to /sbin/grubby
         # which uses the default boot entry in the grub.conf
@@ -1161,7 +1182,11 @@ class Koan:
                 bridge           =  self.virt_bridge,
                 virt_type        =  self.virt_type,
                 virt_auto_boot   =  virt_auto_boot,
-                qemu_driver_type =  self.qemu_disk_type
+                qemu_driver_type  =  self.qemu_disk_type,
+                qemu_net_type     =  self.qemu_net_type,
+                cpu               =  self.virtinstall_cpu,
+                wait              =  self.virtinstall_wait,
+                noreboot          =  self.virtinstall_noreboot
         )
 
         print results
diff -rupN cobbler-2.0.11-old2/koan/virtinstall.py cobbler-2.0.11/koan/virtinstall.py
--- cobbler-2.0.11-old2/koan/virtinstall.py	2013-12-12 17:32:03.506122558 -0500
+++ cobbler-2.0.11/koan/virtinstall.py	2013-12-12 17:59:49.360826677 -0500
@@ -162,6 +162,7 @@ def build_commandline(uri,
                       qemu_driver_type=None,
                       qemu_net_type=None,
                       qemu_machine_type=None,
+                      cpu=None,
                       wait=0,
                       noreboot=False,
                       osimport=False):
@@ -404,6 +405,8 @@ def build_commandline(uri,
         cmd += " "
 
     cmd += "--wait %d " % int(wait)
+    if cpu:
+        cmd += "--cpu %s " % cpu
     if noreboot:
         cmd += "--noreboot "
     if osimport and not(import_exists):
