diff -rupN cobbler-2.0.11.old/cobbler/template_api.py cobbler-2.0.11/cobbler/template_api.py
--- cobbler-2.0.11.old/cobbler/template_api.py
+++ cobbler-2.0.11/cobbler/template_api.py
@@ -314,7 +314,7 @@ class Template(BuiltinTemplate, MacrosTe
                     pass
 
         try:
-            return "#errorCatcher ListErrors\n" + utils.read_file_contents('%s/%s' % (self.getVar('snippetsdir'), file), fetch_if_remote=True)
+            return "#unicode UTF-8\n#errorCatcher ListErrors\n" + utils.read_file_contents('%s/%s' % (self.getVar('snippetsdir'), file), fetch_if_remote=True)
         except FileNotFoundException:
             return None

diff -rupN cobbler-2.0.11.old/cobbler/templar.py cobbler-2.0.11/cobbler/templar.py
--- cobbler-2.0.11.old/cobbler/templar.py
+++ cobbler-2.0.11/cobbler/templar.py 
@@ -117,6 +117,9 @@ class Templar:
         # tell Cheetah not to blow up if it can't find a symbol for something
         raw_data = "#errorCatcher ListErrors\n" + raw_data
 
+        # specify unicode encoding for Cheetah Compiler
+        raw_data = "#unicode UTF-8\n" + raw_data
+
         table_copy = search_table.copy()
 
         # for various reasons we may want to call a module inside a template and pass
