--- cobbler-2.0.11-orig/cobbler/template_api.py	2011-04-20 08:40:48.000000000 -0400
+++ cobbler-2.0.11/cobbler/template_api.py	2014-09-03 09:58:49.947533119 -0400
@@ -63,6 +63,7 @@ DEFAULT_WHITELIST = [
     'silent',
     'slurp',
     'raw',
+    'while',
 ]
 
 WHITELIST = data.get('safe_templating_whitelist', DEFAULT_WHITELIST)
--- cobbler-2.0.11-orig/installer_templates/settings.template 2015-05-11 15:19:03.639582634 +0200
+++ cobbler-2.0.11/installer_templates/settings.template      2015-05-11 15:19:15.429679420 +0200
@@ -414,3 +414,4 @@ safe_templating_whitelist:
    - silent
    - slurp
    - raw
+   - while
