From 70096d0bcc07f91e6f6dd827ea3a179d486bfe92 Mon Sep 17 00:00:00 2001
From: Colin Coe <colin.coe@gmail.com>
Date: Thu, 28 Aug 2008 10:30:24 +0800
Subject: [PATCH] Sort package list in Kickstart profile

---
 .../action/kickstart/EditPackagesAction.java       |   32 +++++++++++++++++++-
 1 files changed, 31 insertions(+), 1 deletions(-)

diff --git a/java/code/src/com/redhat/rhn/frontend/action/kickstart/EditPackagesAction.java b/java/code/src/com/redhat/rhn/frontend/action/kickstart/EditPackagesAction.java
index 6243170..a830c5c 100644
--- a/java/code/src/com/redhat/rhn/frontend/action/kickstart/EditPackagesAction.java
+++ b/java/code/src/com/redhat/rhn/frontend/action/kickstart/EditPackagesAction.java
@@ -27,10 +27,14 @@ import org.apache.struts.action.ActionForward;
 import org.apache.struts.action.ActionMapping;
 import org.apache.struts.action.DynaActionForm;
 
+import java.util.Arrays;
 import java.util.HashSet;
 import java.util.Iterator;
 import java.util.Set;
 import java.util.StringTokenizer;
+import java.io.Reader;
+import java.io.BufferedReader;
+import java.io.StringReader;
 
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
@@ -121,12 +125,38 @@ public class EditPackagesAction extends RhnAction {
         Set packageNames = ksdata.getPackageNames();
         if (packageNames != null && packageNames.size() > 0) {
             StringBuffer buf = new StringBuffer();
+            StringBuffer buf2 = new StringBuffer();
+            int i = 0;
             for (Iterator iter = packageNames.iterator(); iter.hasNext();) {
                 PackageName pn = (PackageName) iter.next();
                 buf.append(pn.getName());
                 buf.append("\n");
+                i++;
             }
-            form.set(PACKAGE_LIST, buf.toString());
+
+            String[] StringArr = new String[i];
+            String line;
+
+            try {
+                BufferedReader in = new BufferedReader(new StringReader(buf.toString()));
+
+                i = 0;
+                while ((line = in.readLine()) != null) {
+                    StringArr[i] = line;
+                    i++;
+                }
+            } catch (Exception e) {
+                e.printStackTrace();
+            }
+
+            Arrays.sort(StringArr);
+
+            for (int j = 0; j < i; j++) {
+                buf2.append(StringArr[j]);
+                buf2.append("\n");
+            }
+
+            form.set(PACKAGE_LIST, buf2.toString());
         }
         form.set("submitted", Boolean.TRUE);
     }
-- 
1.5.5.1

