diff -ru jpam-0.4-orig/src/c/Pam.c jpam-0.4/src/c/Pam.c
--- jpam-0.4-orig/src/c/Pam.c	2012-10-16 17:01:12.000000000 +0200
+++ jpam-0.4/src/c/Pam.c	2012-10-16 17:11:25.079162008 +0200
@@ -156,15 +156,10 @@
 	/* jpam does not yet support password changing, we just copy password */
 	case PAM_PROMPT_ECHO_OFF:
 	case PAM_PROMPT_ECHO_ON:
-		/* Get a prompt, fill the password */
-		if ( 	(! strcmp(msg[replies]->msg, PASS_PROMPT_SECUREID)) ||
-			(! strcmp(msg[replies]->msg, PASS_PROMPT_DEFAULT)) ||
-			(! strcmp(msg[replies]->msg, PASS_PROMPT_MACOSX)) )
-		{
-			if (debug)
-				printf("***Sending password\n");
-			reply[replies].resp = COPY_STRING(password);
-		}
+		/* Ignore the prompt, fill the password */
+		if (debug)
+			printf("***Sending password\n");
+		reply[replies].resp = COPY_STRING(password);
 		break;
 
 	/* XXX Not implemented yet. Just ignore.*/
