From 3e542a0c0979b47a3285962ed1a8339ca9e68385 Mon Sep 17 00:00:00 2001
From: Michael Mraka <michael.mraka@redhat.com>
Date: Wed, 6 Aug 2014 10:58:52 +0200
Subject: [PATCH 1/2] fixed changelog formating

---
 src/tito/rheltagger.py | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/tito/rheltagger.py b/src/tito/rheltagger.py
index fb49fab..9b34de4 100644
--- a/src/tito/rheltagger.py
+++ b/src/tito/rheltagger.py
@@ -27,8 +27,8 @@ class RHELTagger(ReleaseTagger):
         use format:
         - Resolves: #1111 - description
         """
-        patch_command = "git log --pretty='format:%%s%s'" \
-                         " --relative %s..%s -- %s" % (self._changelog_email(), last_tag, "HEAD", ".")
+        patch_command = "git log --pretty='format:%s'" \
+                         " --relative %s..%s -- %s" % (self._changelog_format(), last_tag, "HEAD", ".")
         output = run_command(patch_command)
         BZ = {}
         result = None
--
1.8.3.1
