*** /usr/local/perl5.005_03/lib/site_perl/5.005/HTTP/Daemon.pm.orig	Wed Jun  9 19:41:53 1999
--- /usr/local/perl5.005_03/lib/site_perl/5.005/HTTP/Daemon.pm	Wed Jun  9 19:42:43 1999
***************
*** 63,69 ****
  $VERSION = sprintf("%d.%02d", q$Revision: 1.1.1.1 $ =~ /(\d+)\.(\d+)/);
  
  use IO::Socket ();
! @ISA=qw(IO::Socket::INET);
  
  $PROTO = "HTTP/1.1";
  
--- 63,71 ----
  $VERSION = sprintf("%d.%02d", q$Revision: 1.1.1.1 $ =~ /(\d+)\.(\d+)/);
  
  use IO::Socket ();
! use IO::Socket::SSL;
! #@ISA=qw(IO::Socket::INET);
! @ISA=qw(IO::Socket::SSL);
  
  $PROTO = "HTTP/1.1";
  
***************
*** 156,162 ****
  
  use vars qw(@ISA $DEBUG);
  use IO::Socket ();
! @ISA=qw(IO::Socket::INET);
  *DEBUG = \$HTTP::Daemon::DEBUG;
  
  use HTTP::Request  ();
--- 158,165 ----
  
  use vars qw(@ISA $DEBUG);
  use IO::Socket ();
! #@ISA=qw(IO::Socket::INET);
! @ISA=qw(IO::Socket::SSL);
  *DEBUG = \$HTTP::Daemon::DEBUG;
  
  use HTTP::Request  ();
