*** /usr/local/perl5.005_03/lib/site_perl/5.005/LWP/Protocol/https.pm.orig	Wed Jun  9 19:43:06 1999
--- /usr/local/perl5.005_03/lib/site_perl/5.005/LWP/Protocol/https.pm	Wed Jun  9 19:43:31 1999
***************
*** 4,10 ****
  use strict;
  
  package LWP::Protocol::https;
! require Net::SSL;  # from Crypt-SSLeay
  
  use vars qw(@ISA);
  
--- 4,11 ----
  use strict;
  
  package LWP::Protocol::https;
! #require Net::SSL;  # from Crypt-SSLeay
! require IO::Socket::SSL;
  
  use vars qw(@ISA);
  
***************
*** 15,21 ****
  {
      my($self, $host, $port, $timeout) = @_;
      local($^W) = 0;  # IO::Socket::INET can be noisy
!     my $sock = Net::SSL->new(PeerAddr => $host,
  			     PeerPort => $port,
  			     Proto    => 'tcp',
  			     Timeout  => $timeout,
--- 16,22 ----
  {
      my($self, $host, $port, $timeout) = @_;
      local($^W) = 0;  # IO::Socket::INET can be noisy
!     my $sock = IO::Socket::SSL->new(PeerAddr => $host,
  			     PeerPort => $port,
  			     Proto    => 'tcp',
  			     Timeout  => $timeout,
