diff --git a/client/rhel/dnf-plugin-spacewalk/spacewalk.py b/client/rhel/dnf-plugin-spacewalk/spacewalk.py
index 72ac31e..269f94f 100644
--- a/client/rhel/dnf-plugin-spacewalk/spacewalk.py
+++ b/client/rhel/dnf-plugin-spacewalk/spacewalk.py
@@ -28,7 +28,6 @@ import librepo
 import os
 import sys
 from copy import copy
-from dnf.conf.config import PRIO_PLUGINCONFIG

 # up2date libs are in non-standard path
 sys.path.append("/usr/share/rhn/")
@@ -40,6 +39,7 @@ from rhn.i18n import ustr
 from up2date_client import up2dateErrors
 
 STORED_CHANNELS_NAME = '_spacewalk.json'
+PLUGIN_CONF = 'spacewalk'
 
 RHN_DISABLED    = _("Spacewalk based repositories will be disabled.")
 CHANNELS_DISABLED = _("Spacewalk channel support will be disabled.")
@@ -66,11 +66,11 @@ class Spacewalk(dnf.Plugin):
         self.connected_to_spacewalk = False
         self.up2date_cfg = {}
         self.conf = copy(self.base.conf)
-        self.parser = self.read_config(self.conf)
+        self.parser = self.read_config(self.conf, PLUGIN_CONF)
         if "main" in self.parser.sections():
             options = self.parser.items("main")
             for (key, value) in options:
-                self.conf._set_value(key, value, PRIO_PLUGINCONFIG)
+                setattr(self.conf, key, value)
         if not self.conf.enabled:
             return
         logger.debug('initialized Spacewalk plugin')
@@ -142,9 +142,9 @@ class Spacewalk(dnf.Plugin):
             if channel_id in self.parser.sections():
                 options = self.parser.items(channel_id)
                 for (key, value) in options:
-                    conf._set_value(key, value, PRIO_PLUGINCONFIG)
+                    setattr(conf, key, value)
             repo = SpacewalkRepo(channel_dict, {
-                                    'conf'      : self.base.conf,
+                                    'cachedir'  : self.base.conf.cachedir,
                                     'proxy'     : proxy_url,
                                     'timeout'   : conf.timeout,
                                     'sslcacert' : sslcacert,
@@ -212,7 +212,7 @@ class  SpacewalkRepo(dnf.repo.Repo):
 
     def __init__(self, channel, opts):
         super(SpacewalkRepo, self).__init__(ustr(channel['label']),
-                                            opts.get('conf'))
+                                            opts.get('cachedir'))
         # dnf stuff
         self.name = ustr(channel['name'])
         self.baseurl = [ url + '/GET-REQ/' + self.id for url in channel['url']]
diff --git a/client/rhel/dnf-plugin-spacewalk/actions/packages.py b/client/rhel/dnf-plugin-spacewalk/actions/packages.py
index 86ab72b..5516d1c 100644
--- a/client/rhel/dnf-plugin-spacewalk/actions/packages.py
+++ b/client/rhel/dnf-plugin-spacewalk/actions/packages.py
@@ -271,8 +271,10 @@ def _dnf_base(load_system_repo=True, load_available_repos=True):
     # initialize dnf
     base = dnf.Base()
 
-    if not base.plugins.plugins:
-        base.init_plugins()
+    # this is actually workaround for RhBug:1218071
+    if not base.plugins.plugins and base.conf.plugins:
+        base.plugins.load(base.conf.pluginpath, [])
+        base.plugins.run_init(base)
     if load_available_repos:
         base.read_all_repos()
     base.fill_sack(load_system_repo=True, load_available_repos=True)
