--- PyPAM-0.4.2/PAMmodule.c	1999-08-17 02:27:27.000000000 -0400
+++ PyPAM-0.4.2/PAMmodule.c	2005-10-04 12:48:09.000000000 -0400
@@ -317,6 +317,7 @@
 		result = pam_set_item(_self->pamh, item, (void *) n_val);
 	} else {
 		if (PyArg_ParseTuple(args, "iO:set_callback", &item, &o_val)) {
+			PyErr_Clear();
 			if (item == PAM_CONV && !PyCallable_Check(o_val)) {
 				PyErr_SetString(PyExc_TypeError, "parameter must be a function");
 				return NULL;
