--- PyPAM-0.4.2/PAMmodule.c.orig	2006-12-13 13:52:10.000000000 -0500
+++ PyPAM-0.4.2/PAMmodule.c	2006-12-13 13:53:24.000000000 -0500
@@ -123,8 +123,8 @@
 	p->user = NULL;
 	p->callback = NULL;
 	
-	p->dlh1 = dlopen("libpam.so", RTLD_LAZY | RTLD_GLOBAL);
-	p->dlh2 = dlopen("libpam_misc.so", RTLD_LAZY | RTLD_GLOBAL);
+	p->dlh1 = dlopen("libpam.so.0", RTLD_LAZY | RTLD_GLOBAL);
+	p->dlh2 = dlopen("libpam_misc.so.0", RTLD_LAZY | RTLD_GLOBAL);
 
 	return (PyObject *) p;
 }
