Index: mod_perl-2.0.5/lib/Apache2/Build.pm
===================================================================
--- mod_perl-2.0.5.orig/lib/Apache2/Build.pm	2011-02-02 21:23:47.000000000 +0100
+++ mod_perl-2.0.5/lib/Apache2/Build.pm	2011-05-26 17:12:46.276946520 +0200
@@ -2147,7 +2147,8 @@ sub has_large_files_conflict {
     # with it is that we didn't have such a case yet, but may need to
     # deal with it later
 
-    return $perl_lfs64 ^ $apr_lfs64;
+    return 0;
+    # $perl_lfs64 ^ $apr_lfs64;
 }
 
 # if perl is built with uselargefiles, but apr not, the build won't
