--- Frontier-RPC-0.06/lib/Frontier/RPC2.pm.undef-scalar	2005-06-15 11:28:48.323851492 -0400
+++ Frontier-RPC-0.06/lib/Frontier/RPC2.pm	2005-06-15 12:24:44.739297487 -0400
@@ -226,6 +226,8 @@
 sub _scalar {
     my $self = shift; my $value = shift;
 
+    $value = '' unless defined $value;
+
     # these are from `perldata(1)'
     if ($value =~ /^[+-]?\d+$/) {
 	return ("<value><i4>$value</i4></value>");
