--- GD-1.40/t/GD.t.noxpm	Fri Jul 12 15:28:30 2002
+++ GD-1.40/t/GD.t	Fri Jul 12 15:29:25 2002
@@ -263,6 +263,7 @@
 
 sub test8 {
   my $fn = shift;
+  return 1;
   my $im = GD::Image->newFromXpm($fn);
   $im->png;
 }
--- GD-1.40/Makefile.PL.noxpm	Fri Jul 12 15:24:35 2002
+++ GD-1.40/Makefile.PL	Fri Jul 12 15:24:47 2002
@@ -12,7 +12,7 @@
 
 my $JPEG    = lc prompt('Build JPEG support?','y') eq 'y';
 my $TTF     = lc prompt('Build FreeType support?','y') eq 'y';
-my $XPM     = $^O !~ /^freebsd|MSWin32$/ && lc prompt('Build XPM support?','y') eq 'y';
+my $XPM     = $^O !~ /^freebsd|MSWin32$/ && lc prompt('Build XPM support?','n') eq 'y';
 my $GIF		= lc prompt('Build GIF support (for patched versions of gd)?','n') eq 'y';
 
 my $FCGI    = 0;   # set to 1 to build compatability with fastCGI
