--- Crypt-RIPEMD160-0.04/rmd160.h.unclean	2005-04-18 17:33:37.436753037 -0400
+++ Crypt-RIPEMD160-0.04/rmd160.h	2005-04-18 17:34:08.978647675 -0400
@@ -23,9 +23,11 @@
 /* typedef 8, 16 and 32 bit types, resp.  */
 /* adapt these, if necessary, 
    for your operating system and compiler */
+#include   <inttypes.h>
+
 typedef    unsigned char        byte;   /* unsigned 8-bit integer */
-typedef    unsigned short       word;   /* unsigned 16-bit integer */
-typedef    unsigned long        dword;  /* unsigned 32-bit integer */ 
+typedef    uint16_t       word;   /* unsigned 16-bit integer */
+typedef    uint32_t        dword;  /* unsigned 32-bit integer */ 
 
 
 /********************************************************************/
--- Crypt-RIPEMD160-0.04/rmd160/hash/rmd160.h.unclean	2005-04-18 17:33:57.836451150 -0400
+++ Crypt-RIPEMD160-0.04/rmd160/hash/rmd160.h	2005-04-18 17:34:22.888396242 -0400
@@ -23,9 +23,11 @@
 /* typedef 8, 16 and 32 bit types, resp.  */
 /* adapt these, if necessary, 
    for your operating system and compiler */
+#include   <inttypes.h>
+
 typedef    unsigned char        byte;   /* unsigned 8-bit integer */
-typedef    unsigned short       word;   /* unsigned 16-bit integer */
-typedef    unsigned long        dword;  /* unsigned 32-bit integer */ 
+typedef    uint16_t       word;   /* unsigned 16-bit integer */
+typedef    uint32_t        dword;  /* unsigned 32-bit integer */ 
 
 
 /********************************************************************/
